server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1252,6 +1252,11 @@ private void parseSource(DefaultSearchContext context, SearchSourceBuilder sourc
         QueryBuilder query = source.query();
         if (query != null) {
             InnerHitContextBuilder.extractInnerHits(query, innerHitBuilders);
+            // apply alias filter to as preFilter to queries that need it (e.g. knn)
+            QueryBuilder aliasFilter = context.request().getAliasFilter().getQueryBuilder();
+            if (aliasFilter != null) {
+                query.addFilterQuery(aliasFilter);
+            }
```

Comment:
What do you think about adding the `aliasFilter` to the `SearchExecutionContext` ?

Earlier in this method:
```
SearchExecutionContext searchExecutionContext = context.getSearchExecutionContext();
```

During rewrite the knnQueryBuilder could check `searchExecutionContext#getAliasFilter()` and apply the filters internally.

This seems simpler than adding a new method to all query builder objects (which doesn't seem strictly required to me).

Reply:
That's a great suggestion, the code change is much smaller now. Addressed in 9b1e7c86bea2e72e23b24245cabafe1ff3c823d9